Add 'direction' allow to 'SegmentedButton' oriented vertically

abdalmonem opened the pull request:
This PR add the ability to change buttons of 'SegmentedButton' directionality (In the vertical and horizontal axis) to be 'vertical' or 'horizontal' instead of just horizontally position by adding "direction" argument.

`direction: Axis.horizontal` :
![Simulator Screenshot - iPhone 15 - 2024-06-26 at 13 37 26](https://github.com/flutter/flutter/assets/9139030/4936b7f8-246b-41ae-ac1c-7c75bc2d4f2d)

`direction: Axis.vertical` :
![Simulator Screenshot - iPhone 15 - 2024-06-26 at 13 43 07](https://github.com/flutter/flutter/assets/9139030/5aecf229-34d8-4608-a0f7-aee5c130257f)

Notice: in this example i used:
`style: ButtonStyle( shape: MaterialStateProperty.all<RoundedRectangleBorder>( const RoundedRectangleBorder( borderRadius: BorderRadius.zero, ), ), ) `
To change the Radius of `SegmentedButton`, and the default shape will be like:
![Simulator Screenshot - iPhone 15 - 2024-06-26 at 13 51 46](https://github.com/flutter/flutter/assets/9139030/24833153-02c8-4f5c-8c50-5a0effa19e9e)
I keep it as it is right now, cause its not the main purpose of this BR.



*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*
Fixes: #150416

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

abdalmonem committed:
Add 'direction' allow to 'SegmentedButton' vertically

abdalmonem committed:
Add 'direction' allow to 'SegmentedButton' vertically

abdalmonem committed:
re-init

abdalmonem committed:
re-init

abdalmonem committed:
Initialize PR

abdalmonem committed:
Initialize PR

abdalmonem commented:
https://github.com/flutter/flutter/pull/150854
@Piinks, @QuncCccccc  I opened this new PR and avoided IDE auto format to keep code readable.

abdalmonem committed:
Merge branch 'master' into master

abdalmonem committed:
Merge branch 'master' into master

abdalmonem committed:
Merge branch 'master' into master

abdalmonem committed:
Merge branch 'master' into master

QuncCccccc commented:
Hi @abdalmonem! Thanks a lot for the contribution:) Could you try to rebase master again to see whether the failures can be fixed? I also encountered the same problem and rebasing master works for me.

abdalmonem commented:
Hello @QuncCccccc , yah of course.

abdalmonem committed:
Merge branch 'master' of https://github.com/abdalmonem/150416

abdalmonem commented:
@QuncCccccc  Yeeees all checks have passed 👍

QuncCccccc commented on packages/flutter/test/material/segmented_button_test.dart:
```diff
@@ -1122,6 +1122,44 @@ void main() {
+    );
+
+    final Finder button = find.byType(SegmentedButton<int>);
+    expect(tester.getSize(button).height,
```
I think we might also want to check whether each segmented is aligned vertically with the same left and right, and the bottom of the previous one should be the top of the next one. I think `tester.getRect(find.byType(SegmentedButton<int>))` can be used to get the whole segmented button's size and position and tester.getRect(find.widgetWithText(TextButton, 'Option x')) can be used to check whether each segment inside has correct location and size:)

abdalmonem replied:
Nice tips, am going to work on it.

abdalmonem replied:
@QuncCccccc there is something weird going on when i used `tester.getRect(find.widgetWithText(TextButton, 'Option 0'));` to check the first element ,  the test finder cannot detect  ButtonSegment if its value is zero

abdalmonem replied:
I find out that `TextButton` is removed underneath from (selected) segment button.

abdalmonem replied:
I found a solution , by set `selected` to unknown value in the test, to avoid `TextButton` lost from the elements tree, every thing fine now 👍 .

QuncCccccc reviewed (commented):
Thanks so much for the contribution! Overall this looks great! Just left some suggestions and questions below:)

abdalmonem commented:
@Piinks Yah and i am sorry for the late response , am going to contenu improvement this pr depends on @QuncCccccc review notes

abdalmonem committed:
Update packages/flutter/lib/src/material/segmented_button.dart

Co-authored-by: Qun Cheng <36861262+QuncCccccc@users.noreply.github.com>

abdalmonem committed:
Update packages/flutter/lib/src/material/segmented_button.dart

Oh! my bad :)

Co-authored-by: Qun Cheng <36861262+QuncCccccc@users.noreply.github.com>

abdalmonem commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -564,7 +570,8 @@ class SegmentedButtonState<T> extends State<SegmentedButton<T>> {
             disabledBorder: disabledBorder,
-            direction: direction,
+            direction: widget.direction,
+            textDirection: testDirection,
```
Oh! my bad :)

abdalmonem committed:
Update packages/flutter/lib/src/material/segmented_button.dart

Co-authored-by: Qun Cheng <36861262+QuncCccccc@users.noreply.github.com>

abdalmonem commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -828,7 +862,33 @@ class _RenderSegmentedButton<T> extends RenderBox with
+    RenderBox? child = firstChild;
+    double childHeight;
+    if (_isExpanded) {
+      childHeight = constraints.maxWidth / childCount;
```
I'm currently doing some investigation, i dont remember why i keept it like that even though the variable name is linguistically chaotic.

abdalmonem replied:
After reading documentations of the "expandedInsets" option , i find out that i am wrong and my code lead segmentedButton to miss behave when we set `expandedInsets` option , so am going to commit your suggestions about this.

abdalmonem replied:
"/// Determines the segmented button's size and padding based on [expandedInsets].
  ///
  /// If null (default), the button adopts its intrinsic content size. When specified,
  /// the button expands to fill its parent's space, with the [EdgeInsets]
  /// defining the padding.
  final EdgeInsets? expandedInsets;"

abdalmonem committed:
Update packages/flutter/test/material/segmented_button_test.dart

Co-authored-by: Qun Cheng <36861262+QuncCccccc@users.noreply.github.com>

abdalmonem committed:
Improve test, fix constraints

Improve test by checking all items is vertically aligned using rect dimensions.
Fix expand constraints on the vertical direction.

abdalmonem committed:
Merge branch 'master' into master

abdalmonem commented:
@QuncCccccc @Piinks , its ready for preview.

Piinks commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -154,6 +155,11 @@ class SegmentedButton<T> extends StatefulWidget {
   /// [ChoiceChip] widgets.
   final List<ButtonSegment<T>> segments;
 
+  /// How the [segments] get aligned, by default its [Axis.horizontal].
+  /// If this is [Axis.vertical], the segments will be aligned vertically
+  /// and the first item in [segments] will be on the top.
```
```suggestion
  /// The orientation of the button's [segments].
  ///
  /// If this is [Axis.vertical], the segments will be aligned vertically
  /// and the first item in [segments] will be on the top.
  ///
  /// Defaults to [Axis.horizontal].
```

Piinks commented on packages/flutter/test/material/segmented_button_test.dart:
```diff
@@ -1122,6 +1122,52 @@ void main() {
+      }
+      previewsChildRect = currentChildRect;
+    }
+
```
```suggestion
```

Piinks commented on packages/flutter/test/material/segmented_button_test.dart:
```diff
@@ -1122,6 +1122,52 @@ void main() {
     );
   }, skip: kIsWeb && !isSkiaWeb); // https://github.com/flutter/flutter/issues/99933
+
+  testWidgets('SegmentedButton vertical aligned children', (WidgetTester tester) async {
```
Can you add a golden file test as well? Look for `matchesGoldenFile` in the test directory for an example.

abdalmonem replied:
I added a golden test , but i add it as a separated test , should i make it within the same old test scoop "SegmentedButton vertical aligned children" ?

abdalmonem replied:
"Can you take a look at the analysis failures?"
I see "Linux analyze" some times fail , but i don't think it has something with my PR, its randomly keep fail and success , point me if i am wrong :)
+ Am sorry, i should update with rebase but i always forget.

Piinks reviewed (commented):
Can you take a look at the analysis failures?

Also, should this be added to the theme?

QuncCccccc commented on packages/flutter/test/material/segmented_button_test.dart:
```diff
@@ -1122,6 +1122,52 @@ void main() {
+    );
+
+    Rect? previewsChildRect;
+    for( int i=0; i <= 3; i++) {
```
```suggestion
    for ( int i=0; i <= 3; i++) {
```

QuncCccccc commented on packages/flutter/test/material/segmented_button_test.dart:
```diff
@@ -1122,6 +1122,52 @@ void main() {
+    Rect? previewsChildRect;
+    for( int i=0; i <= 3; i++) {
+      final Rect currentChildRect = tester.getRect(find.widgetWithText(TextButton, 'Option $i'));
+      if(previewsChildRect != null) {
```
```suggestion
      if (previewsChildRect != null) {
```

abdalmonem committed:
Update packages/flutter/lib/src/material/segmented_button.dart

Co-authored-by: Kate Lovett <katelovett@google.com>

abdalmonem committed:
Update packages/flutter/test/material/segmented_button_test.dart

Co-authored-by: Kate Lovett <katelovett@google.com>

abdalmonem commented:
@Piinks I am not sure about "Also, should this be added to the theme?"

abdalmonem committed:
Update packages/flutter/test/material/segmented_button_test.dart

Co-authored-by: Qun Cheng <36861262+QuncCccccc@users.noreply.github.com>

abdalmonem committed:
Update packages/flutter/test/material/segmented_button_test.dart

Co-authored-by: Qun Cheng <36861262+QuncCccccc@users.noreply.github.com>

abdalmonem committed:
Add golden test to the vertical aligned segmented button

abdalmonem committed:
Merge branch 'master' into master

QuncCccccc reviewed (commented):
Linux analyzer failed because of this error:
![Screenshot 2024-07-30 at 5 55 27 PM](https://github.com/user-attachments/assets/779d96df-feb3-4889-83f8-620806e7d2f2)

Based on the error message, I think adding `@Tags(<String>['reduced-test-set'])` on top of the test file can fix the issue:)

QuncCccccc commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -835,7 +872,33 @@ class _RenderSegmentedButton<T> extends RenderBox with
+        childHeight = math.max(childHeight, child.getMaxIntrinsicHeight(double.infinity));
+        child = childAfter(child);
+      }
+      childHeight = math.min(childHeight, constraints.maxWidth / childCount);
```
I think this should also be `constraints.maxHeight / childCount`? Because if direction is vertical, we should consider the height of the constraints instead of width, right?

abdalmonem replied:
I can see my code can lead to this bad behavior :) when the parent height is less than the `segmentedButton` height , and can lead to hide some items in the list , now i can confirm this, thanks Qun.

QuncCccccc commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -835,7 +872,33 @@ class _RenderSegmentedButton<T> extends RenderBox with
+    }
+    child = firstChild;
+    while (child != null) {
+      final double boxHeight = child.getMaxIntrinsicWidth(maxWidth);
```
This variable name is a little confusing. Would be better if we can change it to `boxWidth`:)

abdalmonem replied:
I agree :)

QuncCccccc commented on packages/flutter/lib/src/material/segmented_button.dart:
```diff
@@ -941,9 +1004,15 @@ class _RenderSegmentedButton<T> extends RenderBox with
+          final Offset bottom = Offset(dividerPos, borderRect.bottom);
+          context.canvas.drawLine(top, bottom, divider.toPaint());
+        } else if (direction == Axis.vertical) {
+          final Offset start = Offset(borderRect.left, childRect.top);
+          final Offset end = Offset(borderRect.right, childRect.top);
```
Should both start.y and end.y be `childRect.bottom`?

abdalmonem replied:
Yup , this part is responsible for divisions lines, and currently its correct.
Its get draw by the the start of (topLeft of the (element.index-1)) to (topRight of the (element.index-1))
Did i get you right ?

QuncCccccc replied:
I see. I think we also need to consider to clip the divider here. It seems when it's a rounded rectangle segmented buttons, the divider might exceed the border.

<img width="300" src="https://github.com/user-attachments/assets/4a943048-bf27-468e-9a31-062a123fb4da">

abdalmonem replied:
hi @QuncCccccc , do you have any suggestion to do it ?

abdalmonem replied:
Yeeeeees i solve it @QuncCccccc  , I check out flutter team code pattern , and i figure out how to do it, you can check it, let me know if you have any feedback's.
![Simulator Screenshot - iPhone 15 - 2024-09-05 at 14 53 21](https://github.com/user-attachments/assets/29377166-4d14-4b42-958e-9e4e5d081af6)

QuncCccccc replied:
NICE! Taking a look:)

QuncCccccc reviewed (commented):
Thanks for the update! I just noticed if direction is vertical, with the default shape, the dividers are not clipped correctly as shown in your example. I think it would be better if we can clip the extra parts of the dividers.

Piinks commented:
Hey @abdalmonem, would you like to continue working on this change?

abdalmonem commented:
@Piinks Yes, i will check all @QuncCccccc comments this week.

abdalmonem commented:
@QuncCccccc Am too sorry for my delay, i well check all your comments ASAP.

abdalmonem committed:
Fix segmented buttons get hide due to incorrect child height calculations

abdalmonem committed:
refactoring

Piinks commented:
Hey @abdalmonem can you rebase with the tip of tree to resolve the failing ci.yaml check? Thanks!

abdalmonem committed:
Merge branch 'master' into master

abdalmonem commented:
@Piinks i see "ci.yaml"test got passed successfully.

abdalmonem committed:
Clip extra divider

QuncCccccc reviewed (approved):
LGTM! Thanks for the contribution!

QuncCccccc commented on packages/flutter/test/material/segmented_button_test.dart:
```diff
@@ -1122,6 +1124,86 @@ void main() {
+                    label: Text('Option 1'),
+                  ),
+                ],
+                selected: const <int>{0}, // Prevent any of ButtonSegment to be selected
```
This comment seems unnecessary?

abdalmonem replied:
I agree with u 👍

Piinks reviewed (approved):
LGTM, thank you!
